I have a patch to fix building on Darwin: [0001-dosbox-x-Enable-on-Darwin.patch.txt](https://github.com/NixOS/nixpkgs/files/12795348/0001-dosbox-x-Enable-on-Darwin.patch.txt)

But I can't run the resulting binary on my 10.13 system:
```
dyld: Symbol not found: __objc_empty_cache
  Referenced from: /nix/store/09lrr2yf432m52ihv4m4gz4d27w2d4lw-dosbox-x-2023.09.01/bin/dosbox-x
  Expected in: /System/Library/Frameworks/Carbon.framework/Versions/A/Carbon
 in /nix/store/09lrr2yf432m52ihv4m4gz4d27w2d4lw-dosbox-x-2023.09.01/bin/dosbox-x
```

This *might* be because building this needs the 11.0 SDK for [`futimens`](https://github.com/joncampbell123/dosbox-x/blob/a641673b16099d0ba0fedb64da311c12a0b1aced/src/dos/drive_local.cpp#L2859-L2867), and newer versions of `Carbon` might pull in `libobjc` while my system-shipped one doesn't? I'm not sure.

I tried to change the `#ifdef` there so `futimens` isn't needed and the regular SDK can be used, but that gives a compile error that I don't understand:
```
clang++ -DHAVE_CONFIG_H -I. -I../..  -I../../include -I../../src/aviwriter "-DRESDIR=\"\"" -I../../src -I../../src/libs/mt32 -Wno-int-to-void-pointer-cast   -Wno-address-of-packed-member   -Wno-format-zero-length   -Wno-missing-field-initializers   -Wno-strict-aliasing   -Wno-implicit-fallthrough   -Wno-deprecated-declarations   -Wconversion-null   -Wsign-promo   -pedantic   -Wunused   -Wextra   -Wall   -I/usr/local/include -D_THREAD_SAFE -I/nix/store/4p0574y0y85qiw5m6pg826x3p5dr9ic6-SDL2-2.28.3-dev/include -I/nix/store/4p0574y0y85qiw5m6pg826x3p5dr9ic6-SDL2-2.28.3-dev/include/SDL2 -I/private/tmp/nix-build-dosbox-x-2023.09.01.drv-0/source -I/private/tmp/nix-build-dosbox-x-2023.09.01.drv-0/source/vs/sdlnet/linux-host/include -I/private/tmp/nix-build-dosbox-x-2023.09.01.drv-0/source/vs/sdlnet/linux-host/include/SDL -I/nix/store/q1x1h7n9d3yycf2j8z5q9aad96n2n5yg-freetype-2.13.1-dev/include/freetype2 -I/nix/store/s05ika6ljwa38c1r08l67dyk46sd24ih-glib-2.76.4-dev/include/glib-2.0 -I/nix/store/ag37hpnacgwm4ci76hdr38jkdl17yrdb-glib-2.76.4/lib/glib-2.0/include -I/nix/store/y28n2cs6v575ia8bn279pqzc8wqjmn9v-libslirp-4.7.0/include/slirp -I/nix/store/6np6h2dnkhjl08b61wybs1yn3m17kddz-ffmpeg-5.1.3-dev/include  -g -std=gnu++14  -O2 -msse  -Wall   -Wextra   -Wunused   -pedantic   -Wsign-promo   -Wconversion-null   -Wno-deprecated-declarations   -Wno-implicit-fallthrough   -Wno-strict-aliasing   -Wno-missing-field-initializers   -Wno-format-zero-length   -Wno-address-of-packed-member   -Wno-int-to-void-pointer-cast  -I/private/tmp/nix-build-dosbox-x-2023.09.01.drv-0/source -I/private/tmp/nix-build-dosbox-x-2023.09.01.drv-0/source/vs/sdlnet/linux-host/include -I/private/tmp/nix-build-dosbox-x-2023.09.01.drv-0/source/vs/sdlnet/linux-host/include/SDL  -c -o sdl_mapper.o sdl_mapper.cpp
sdl_mapper.cpp:5472:61: error: expected unqualified-id
    mainMenu.get_item("hostkey_mapper").check(hostkeyalt==0).set_text("Mapper-defined: "+mapper_keybind).refresh_item(mainMenu);
                                                            ^
1 error generated.
make[3]: *** [Makefile:442: sdl_mapper.o] Error 1
make[3]: Leaving directory '/private/tmp/nix-build-dosbox-x-2023.09.01.drv-0/source/src/gui'
```

---

Could you apply the patch file above + the following in a separate commit and let the OfBorg builder test if the produced binary works on Darwin? If the x86_64-darwin run manages to build `dosbox-x` but happens to have the same issue in `dosbox-x.passthru.tests` that I had, then just revert the commit & we'll consider Darwin support beyond our capabilities for now (unless someone with a newer Darwin box wants to figure this out).

```nix
  passthru.tests.version = testers.testVersion {
    package = finalAttrs.finalPackage;
    # Version output on stderr, program returns status code 1
    command = "${finalAttrs.meta.mainProgram} -version 2>&1 || true";
  }; 
```

--- a/pkgs/applications/emulators/dosbox-x/default.nix
+++ b/pkgs/applications/emulators/dosbox-x/default.nix
@@ -1,0 +1,101 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, alsa-lib
+, AudioUnit
+, autoreconfHook
+, Carbon
+, Cocoa
+, ffmpeg
+, fluidsynth
+, freetype
+, glib
+, libpcap
+, libpng
+, libslirp
+, libxkbfile
+, libXrandr
+, makeWrapper
+, ncurses
+, pkg-config
+, SDL2
+, SDL2_net
+, testers
+, yad
+, zlib
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "dosbox-x";
+  version = "2023.10.06";
+
+  src = fetchFromGitHub {
+    owner = "joncampbell123";
+    repo = "dosbox-x";
+    rev = "dosbox-x-v${finalAttrs.version}";
+    hash = "sha256-YNYtYqcpTOx4xS/LXI53h3S+na8JVpn4w8Dhf4fWNBQ=";
+  };
+
+  strictDeps = true;
+
+  nativeBuildInputs = [
+    autoreconfHook
+    makeWrapper
+    pkg-config
+  ];
+
+  buildInputs = [
+    ffmpeg
+    fluidsynth
+    freetype
+    glib
+    libpcap
+    libpng
+    libslirp
+    ncurses
+    SDL2
+    SDL2_net
+    zlib
+  ] ++ lib.optionals stdenv.hostPlatform.isLinux [
+    alsa-lib
+    libxkbfile
+    libXrandr
+  ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    AudioUnit
+    Carbon
+    Cocoa
+  ];
+
+  configureFlags = [ "--enable-sdl2" ];
+
+  enableParallelBuilding = true;
+
+  hardeningDisable = [ "format" ]; # https://github.com/joncampbell123/dosbox-x/issues/4436
+
+  postInstall = lib.optionalString stdenv.hostPlatform.isLinux ''
+    wrapProgram $out/bin/dosbox-x \
+      --prefix PATH : ${lib.makeBinPath [ yad ]}
+  '';
+
+  passthru.tests.version = testers.testVersion {
+    package = finalAttrs.finalPackage;
+    # Version output on stderr, program returns status code 1
+    command = "${lib.getExe finalAttrs.finalPackage} -version 2>&1 || true";
+  };
+
+  meta = {
+    homepage = "https://dosbox-x.com";
+    description = "A cross-platform DOS emulator based on the DOSBox project";
+    longDescription = ''
+      DOSBox-X is an expanded fork of DOSBox with specific focus on running
+      Windows 3.x/9x/Me, PC-98 and 3D support via 3dfx.
+
+      The full expanded feature list is available here:
+      https://dosbox-x.com/wiki/DOSBox%E2%80%90X%E2%80%99s-Feature-Highlights
+    '';
+    license = lib.licenses.gpl2Plus;
+    maintainers = with lib.maintainers; [ hughobrien OPNA2608 ];
+    platforms = lib.platforms.unix;
+    mainProgram = "dosbox-x";
+  };
+})

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -2614,6 +2614,10 @@ with pkgs;
 
   dosbox-staging = callPackage ../applications/emulators/dosbox-staging { };
 
+  dosbox-x = darwin.apple_sdk_11_0.callPackage ../applications/emulators/dosbox-x {
+    inherit (darwin.apple_sdk_11_0.frameworks) AudioUnit Carbon Cocoa;
+  };
+
   duckstation = qt6Packages.callPackage ../applications/emulators/duckstation { };
 
   dynamips = callPackage ../applications/emulators/dynamips { };

